### ISSUE TITLE
Perl 5.8 + 5.10 support

### ISSUE THREAD

**Issue by plicease** [p5pclub/ref-util#2]
The way Regexp references appears to have changed in 5.8 and then again in 5.12, which is causing failures that can be seen on the cpantesters matrix.

To get it to work on 5.8/5.10 I was trying to adjust the `is_regexpref` function, and cribbing off of Data::Dumper:

https://metacpan.org/source/SMUELLER/Data-Dumper-2.154/Dumper.xs#L481
https://metacpan.org/source/SMUELLER/Data-Dumper-2.154/Dumper.xs#L564

And I was able to get it to compile and pass all the non-regexp tests, but either my logic was flawed or I was misreading the code.  Maybe both. `SvOBJECT(SvRV(ref))` was not returning true for me.

Looks like Perl 5.6 support should also be possible, though that will matter to others more than me :)

**Comment by plicease** [comment]
This works on 5.10, but not 5.8:

``` diff
diff --git a/Util.xs b/Util.xs
index ff78cee..c51bbb9 100644
--- a/Util.xs
+++ b/Util.xs
@@ -135,7 +135,13 @@ is_coderef(SV *ref)
 SV *
 is_regexpref(SV *ref)
     PPCODE:
+#if PERL_VERSION < 11
+        SvRXOK(ref) ? XSRETURN_YES : XSRETURN_NO;
+#else
         XSUB_BODY( ref, ==, SVt_REGEXP );
+#endif

 SV *
 is_globref(SV *ref)
```

**Comment by plicease** [comment]
@karenetheridge suggested pure-perl:

``` perl
re->can('is_regexp') ? re::is_regexp($_[0]) : ref($_[0]) eq 'Regexp'
```

Will work on 5.10.x:

```
twin% perl -e 'my $obj = bless qr{}, "Foo"; print re::is_regexp($obj), "\n"'
1
```

but not 5.8.x for blessed references:

```
twin% perl -e 'my $obj = bless qr{}, "Foo"; print ref $obj, "\n"'
Foo
```

Also `SvRXOK` was introduced in 5.10.1 so even my suggestion will not even help for 5.10.0.

**Comment by xsawyerx** [comment]
I'm sorry I haven't replied to this yet - holidays. I'm going to play with it tomorrow, possibly today. Thanks for raising the issue and tracking all the relevant information!

**Comment by plicease** [comment]
@xsawyerx np get to it when you get to it.

**Comment by xsawyerx** [comment]
I just pushed something which seems to be working on a `perlbrew`ed 5.8.9.

I'm going to try and run it on 5.6.2. Just struggling with dependencies of the build system.

**Comment by xsawyerx** [comment]
5.6.2 works with `is_regexpref`, but fails with `is_formatref`.

**Comment by xsawyerx** [comment]
Seems like `*main::STDOUT{'FORMAT'}` on 5.6.2 does not create a proper `FORMAT` reference. Argh.

So it's a failing test.

I'm not sure how to create a FORMAT reference in that Perl version. Any ideas?

**Comment by xsawyerx** [comment]
Meanwhile added support for 5.6.2+, except for `is_formatref`. The new `is_regexpref` works on it, so it's supported too.

Will play with it a bit more and carve a release tonight/tomorrow.

**Comment by plicease** [comment]
Looks good I tested this on 5.8.8 and 5.10.0 also.  I think you will need to bump the `Test::More` requirement though, this is what I got with vanilla 5.10.0:

```
twin% perl -MTest::More -E 'say Test::More->VERSION'
0.72
twin% perl -E 'say $]'
5.010000
twin% make test
PERL_DL_NONLAZY=1 /home/ollisg/perl5/perlbrew/perls/perl-5.10.0/bin/perl "-MExtUtils::Command::MM" "-e" "test_harness(0, 'blib/lib', 'blib/arch')" t/*.t
t/arrayref.....You tried to run a test without a plan at t/arrayref.t line 10.
# Looks like your test died before it could output anything.
t/arrayref.....dubious                                                       
    Test returned status 255 (wstat 65280, 0xff00)
t/dynamic......ok                                                            
t/expr.........ok                                                            
t/functions....Bareword "done_testing" not allowed while "strict subs" in use at t/functions.t line 36.
Execution of t/functions.t aborted due to compilation errors.
# Looks like your test died before it could output anything.
t/functions....dubious                                                       
    Test returned status 255 (wstat 65280, 0xff00)
Failed Test   Stat Wstat Total Fail  List of Failed
-------------------------------------------------------------------------------
t/arrayref.t   255 65280    ??   ??  ??
t/functions.t  255 65280    ??   ??  ??
Failed 2/4 test scripts. 0/3 subtests failed.
Files=4, Tests=3,  0 wallclock secs ( 0.05 cusr +  0.00 csys =  0.05 CPU)
Failed 2/4 test programs. 0/3 subtests failed.
Makefile:914: recipe for target 'test_dynamic' failed
make: *** [test_dynamic] Error 255
```

**Comment by xsawyerx** [comment]
I don't mind changing the `done_testing` to simple `is()` to support older versions of `Test::More`. That's probably easier than forcing users to upgrade `Test::More` when they run old versions of Perl, no? I'll do that then.

**Comment by plicease** [comment]
yup.

**Comment by xsawyerx** [comment]
With latest release, I consider this resolved.

After checking with p5p, format references are not available in 5.6.

Thanks, @plicease, @karenetheridge!
